Paper for JOSS (#200)
mind-inria/mri-nufft: docs/paper-joss/paper.md

---
title: 'MRI-NUFFT: Doing non-Cartesian MRI has never been easier'
tags:
  - Python
  - MRI
  - NUFFT
  - Numpy 
  - CUDA
  - Torch 
authors:
  - name: Pierre-Antoine Comby
    orcid: 0000-0001-6998-232X
    corresponding: true
    affiliation: "1, 2"

>   - name: Guillaume Daval-Frérot

Reviewer: ```suggestion
  - name: Guillaume Daval-Frérot
    orcid: 0000-0002-5317-2522
```

    affiliation: 3
  - name: Caini Pan 
    affiliation: "1, 2"
  - name: Asma Tanabene
    affiliation: "1,2,5" 
  - name: Léna Oudjman
    affiliation: 1,2
  - name: Matteo Cencini
    affiliation: 4

>   - name: Philippe Ciuciu

Reviewer: ```suggestion
  - name: Philippe Ciuciu
     orcid: 0000-0001-5374-962X
```

    affiliation: "1,2"

>   - name: Chaithya GR

Reviewer: ```suggestion
  - name: Chaithya GR
    orcid: 0000-0001-9859-6006
```

    corresponding: true 
    affiliation: "1,2"
    
affiliations:
 - name: MIND, Inria
   index: 1
 - name: Université Paris-Saclay / CEA 
   index: 2
 - name: Chipiron
   index: 3
 - name: INFN, Pisa Division 
   index: 4 
 - name: Siemens Healthineers 
   index: 5
   
date: 20 September 2024
bibliography: paper.bib
---

# Summary 

> MRI-NUFFT is a python package that provides a universal interface to various Non-Uniform Fast Fourier Transforms libraries running on CPU or GPU (gpunufft, finufft, cufinufft, pynfft), adding compatibily with standard array library (numpy, cupy, torch, tensorflow, etc.) On top of these libraries it extends the existing NUFFT operations to provide a physical model of the MRI acquisition process (e.g. multi-coil acquisition and static-field inhomogeneities). Moreover, it also provides a wide variety of non-Cartesian Sampling trajectories generation and expansion, as well as density-compensation estimation methods for those trajectories. It also implements optimized auto-differentiation with respect to the data and the sampling locations. With MRI-NUFFT one can experiment with non-cartesian sampling in MRI, get access to the latest advances in the field and state-of-the art sampling patterns.

Reviewer 2: I am still not sure because of the logo and the varying names, but is it MRI-_nufft_, or MRI-NUFFT ?

```suggestion
MRI-NUFFT is a python package that provides a universal interface to various Non-Uniform Fast Fourier Transform libraries running on CPU or GPU (gpuNUFFT, FINUFFT, CUFINUFFT, pyNFFT), adding compatibily with standard array library (NumPy, CuPy, PyTorch, TensorFlow, etc.) On top of these libraries it extends the existing NUFFT operations to provide a physical model of the MRI acquisition process (e.g. multi-coil acquisition and static-field inhomogeneities). It also provides a wide variety of customizable implementations of non-Cartesian sampling trajectories, as well as density compensation methods. Finally, it proposes optimized auto-differentiation with respect to the data and sampling locations for machine learning. With MRI-NUFFT one can experiment with non-Cartesian sampling in MRI, get access to the latest advances in the field and state-of-the-art sampling patterns.
```

Author: I have settle with MRI-NUFFT in all caps everywhere, except in the logo...

> # Statement of Need 

Reviewer 2: ```suggestion
# Statement of need 
```

> MRI is an non-invasive biomedical imaging technique, where raw data is sampled in the spatial frequency domain (k-space) and final images are  obtained by applying a (fast) fourier transform on this data.

Reviewer 2: ```suggestion
MRI is an non-invasive biomedical imaging technique, where raw data is sampled in the spatial frequency domain (k-space) and final images are  obtained by applying an inverse (fast) Fourier transform on this data.
```

> Traditionnaly, the data is sampled on a Cartesian grid, potentially with skipping lines (to accelerate the acquisition)  and reconstructed using FFT-based algorithms. 

Reviewer 2: ```suggestion
Traditionnaly, the data is sampled on a Cartesian grid (often partially by skipping lines to accelerate the acquisition)  and reconstructed using FFT-based algorithms. 
```

> However, the Cartesian approach is not always the best choice for collecting the data, and non-Cartesian sampling schemes have been proposed to improve the image quality, reduce the acquisition time or to enable new imaging modalities. But the reconstruction of non-Cartesian data is more challenging and requires the use of non-uniform fast Fourier transform (NUFFT) algorithms. 

Reviewer 2: ```suggestion
However, the Cartesian approach is not always the best choice for data collection, and non-Cartesian sampling schemes have been proposed to improve image quality, reduce acquisition time or enable new imaging modalities. But the reconstruction of non-Cartesian data is more challenging and requires the use of non-uniform fast Fourier transform (NUFFT) algorithms. 
```

> However, the Cartesian grid is not always the best choice for sampling the data, and non-cartesian sampling schemes have been proposed to improve the image quality, reduce the acquisition time or to enable new imaging modalities. The reconstruction of non-cartesian data is more complex than the cartesian case, and requires the use of non-uniform fast Fourier transform (NUFFT) algorithms. 

Reviewer 2: "enable new imaging modalities" ???

```suggestion
However, the Cartesian approach is not always the best choice for collecting the data, and non-Cartesian sampling schemes have been proposed to improve the image quality, reduce the acquisition time or to enable new imaging modalities. But the reconstruction of non-Cartesian data is more challenging and requires the use of non-uniform fast Fourier transform (NUFFT) algorithms. 
```

> Several NUFFT libraries have been developed in the past years, but they are not always easy to use or don't account for the specificities of MRI data acquisition (e.g. multi-coil acquisition, static-field inhomogeneities, density compensation, etc.). Also their performances can vary a lot depending on the use cases (2D vs 3D data, number of coils, etc.). 

Reviewer 2: ```suggestion
Several NUFFT libraries have been developed in the past few years, but they are not always easy to use or don't account for the specificities of MRI data acquisition (e.g. multi-coil acquisition, static-field inhomogeneities, density compensation, etc.). Also their performances can vary a lot depending on the use cases (2D vs 3D data, number of coils, etc.). 
```

> Several NUFFT libraries have been developed in the past years, but they are not always easy to use or implementing the specificity of MRI data acquisition (e.g. multi-coil acquisition, static-field inhomogeneities, density compensation, etc.). Also their performances can vary a lot depending on the specific use-cases (2D vs 3D data, number of coils, etc.). 

Reviewer 2: ```suggestion
Several NUFFT libraries have been developed in the past years, but they are not always easy to use or don't account for the specificities of MRI data acquisition (e.g. multi-coil acquisition, static-field inhomogeneities, density compensation, etc.). Also their performances can vary a lot depending on the use cases (2D vs 3D data, number of coils, etc.). 
```

> Moreover, the use of non-cartesian sampling in MRI is still an active research field, with new sampling patterns being proposed regularly. With MRI-NUFFT one can  easily experiment with these new patterns, and compare them with existing one.

Reviewer 2: ```suggestion
Moreover, non-Cartesian acquisitions are still an active research field, with new sampling patterns being proposed regularly. With MRI-NUFFT one can easily experiment with these new patterns and compare them to existing ones.
```

> Furthermore, there has been a growing interest in using deep learning for MRI acquisition and reconstruction, and using those new methods for Non-Cartesian Data requires to be able to compute the gradients of the reconstruction with respect to the data and/or the sampling locations.

Reviewer 2: ```suggestion
Furthermore, there has been a growing interest in using deep learning to jointly learn MRI acquisition and reconstruction, which requires to compute the gradients of the reconstruction with respect to the raw data and/or the sampling locations.
```

# Features 

> ![MRI-NUFFT as an interface for Non Cartesian MRI](../_static/mri-nufft-scheme.svg)

Reviewer 2: ```suggestion
![MRI-NUFFT as an interface for non-Cartesian MRI](../_static/mri-nufft-scheme.svg)
```

## NUFFT Library compatibility 

> MRI-NUFFT is compatible with the following NUFFT librairies: finufft[@barnett_parallel_2019], cufinufft[@shih_cufinufft_2021], gpunufft[@knoll_gpunufft_2014], torchkbnufft[@muckley_torchkbnufft_2020], pynfft, sigpy[@ong_frank_sigpy_2019] and BART[@uecker_berkley_2015]. 

Reviewer 2: ```suggestion
MRI-NUFFT is compatible with the following NUFFT librairies: FINUFFT[@barnett_parallel_2019], CUFINUFFT[@shih_cufinufft_2021], gpuNUFFT[@knoll_gpunufft_2014], TorchKbNufft[@muckley_torchkbnufft_2020], pyNFFT, sigpy[@ong_frank_sigpy_2019] and BART[@uecker_berkley_2015]. 
```
Capitalizing library names based on their own websites

> Using our [benchmark](https://github.com/mind-inria/mri-nufft-benchmark/) we can also determine which NUFFT implementation provides the best performances both in term of computation time and memory footprint. As the time of writing, cufinufft and gpunufft provide the best performances by leveraging CUDA acceleration. MRI-NUFFT supports as well standard array libraries (numpy, cupy, torch, tensorflow, etc.) and optimizes data copies, relying on the array-API standard. 

Reviewer 2: ```suggestion
Using our [benchmark](https://github.com/mind-inria/mri-nufft-benchmark/) we can also determine which NUFFT implementation provides the best performances both in term of computation time and memory footprint. At the time of writing, cufinufft and gpunufft provide the best performances by leveraging CUDA acceleration. MRI-NUFFT supports as well standard array libraries (NumPy, CuPy, PyTorch, TensorFlow, etc.) and optimizes data copies, relying on the array-API standard. 
```

> Using our [benchmark](https://github.com/mind-inria/mri-nufft-benchmark/) we can also determined which implementations of the NUFFT provides the best performances (both in term of computation time and memory footprint). As the time of writing cufinufft and gpunufft provides the best performances, by leveraging CUDA acceleration. MRI-NUFFT supports as well  standard array libraries (numpy, cupy, torch, tensorflow, etc.) and optimizes data copies, relying on the array-api standard. 

Reviewer 2: ```suggestion
Using our [benchmark](https://github.com/mind-inria/mri-nufft-benchmark/) we can also determine which NUFFT implementation provides the best performances both in term of computation time and memory footprint. As the time of writing, cufinufft and gpunufft provide the best performances by leveraging CUDA acceleration. MRI-NUFFT supports as well standard array libraries (numpy, cupy, torch, tensorflow, etc.) and optimizes data copies, relying on the array-API standard. 
```

> On top of these NUFFT backend, it provides several enhancements, notably an optimized 2.5D NUFFT (for stack of 2D non uniform trajectory, a common pattern in MRI), and a data-consistency term for iterative reeconstruction ($\mathcal{F}_\Omega^*(\mathcal{F}_\Omega x - y)$) that can be used in iterative reconstruction algorithms.

Reviewer 2: ```suggestion
It also provides several enhancements on top of these backends, notably an optimized 2.5D NUFFT (for stacks of 2D non uniform trajectories, commonly used in MRI), and a data-consistency term for iterative reconstruction ($\mathcal{F}_\Omega^*(\mathcal{F}_\Omega x - y)$).
```

Fortunately, this inhomogeneity map can be acquired separately or estimated and integrated in the model as:

$$y(t_i) = \int_{\mathbb{R}^d} x(\boldsymbol{u}) e^{-2\imath\pi \boldsymbol{u} \cdot\boldsymbol{\nu_i} + \Delta\omega(\boldsymbol{u}) t_i} d\boldsymbol{u}$$

where $t_i$ is the time at which the frequency $\nu_i$ is acquired.
With these mixed-domain field pertubations, the Fourier model does not hold anymore and the FFT algorithm cannot be used any longer. 
The main approach [@sutton_fast_2003] is to approximate the mixed-domain exponential term by splitting it into single-domain weights $b_{m, \ell}$ and $c_{\ell, n}, where $L \ll M, N$ regular Fourier transforms are performed to approximate the non-Fourier transform.

$$x(\boldsymbol{u_n}) = \sum_{\ell=1}^L c_{\ell, n} \sum_{m}^M y(t_m) b_{m, \ell} e^{2\imath\pi \boldsymbol{u} \cdot \boldsymbol{\nu_i}}$$

The coefficients $B=(b_{m, \ell}) \in \mathbb{C}^{M\times L}$ and $C=(c_\ell, n) \in \mathbb{C}^{L\times N}$ can be optimally estimated within MRI-NUFFT.

## Trajectories generation and expansions 

> MRI-NUFFT comes with a wide variety of Non Cartesian trajectory generation routines, that have been gathered from the literature. It also provides ways of expanding existing trajectories. It is also able to export to specific formats, to be used in other toolboxes and on MRI hardware.

Reviewer: Maybe we add links to docs? Is it allowed?

Author: No need the docs is reviewed along the paper itself. and the repo is linked in it already

> MRI-NUFFT comes with a wide variety of Non Cartesian trajectory generation routines, that have been gathered from the literature. It also provides ways of expanding existing trajectories. It is also able to export to specific formats, to be used in other toolboxes and on MRI hardware.

Reviewer 2: ```suggestion
MRI-NUFFT comes with a wide variety of non-Cartesian trajectory generation routines that have been gathered from the literature. It also provides ways to extend existing trajectories and export them to specific formats, for use in other toolkits and on MRI hardware.
```

> ## Autodifferentiation for data and sampling pattern

Reviewer 2: ```suggestion
## Auto-differentiation for data and sampling pattern
```

> Following the formulation of [@wang_efficient_2023], MRI-NUFFT also provides autodifferentation capabilities for all the NUFFT backends. Both gradients with respect to the data (image or kspace) and the sampling point location are available. This allows for efficient backpropagation throught the NUFFT operators, and sustains research on learned sampling pattern and image reconstruction network.

Reviewer 2: ```suggestion
Following the formulation of [@wang_efficient_2023], MRI-NUFFT provides automatic differentiation for all NUFFT backends, with respect to both gradients and data (image or k-space). This enables efficient backpropagation through NUFFT operators and supports research on learned sampling model and image reconstruction network.
```

> # MRI-NUFFT Utilization

Reviewer 2: ```suggestion
# MRI-NUFFT utilization
```
